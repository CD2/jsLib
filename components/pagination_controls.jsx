import React from 'react'
import PropTypes from 'prop-types'
import decorate from 'utils/decorate'
import { styled, t } from 'utils/theme'


export class PaginationControls extends React.Component {

  static propTypes = {
    className: PropTypes.string,
    onPageChange: PropTypes.func.isRequired,
    page: PropTypes.number.isRequired,
    per_page: PropTypes.number.isRequired,
    total_records: PropTypes.number.isRequired,
  }

  changePage(page_number) {
    this.props.onPageChange(page_number)
  }

  totalPages() {
    const { total_records, per_page } = this.props
    return Math.ceil(total_records/per_page)
  }

  firstPage() {
    return this.props.page === 1
  }

  lastPage() {
    return this.props.page === this.totalPages()
  }

  renderLeft() {
    const  className = `pagination__button`
    if (this.firstPage()) {
      return (
        <div className={`${className} disabled`}>Left</div>
      )
    }
    return (
      <div className={className} onClick={() => this.changePage(this.props.page - 1)}>Left</div>
    )
  }

  renderRight() {
    const  className = `pagination__button`
    if (this.lastPage()) {
      return (
        <div className={`${className} disabled`}>Right</div>
      )
    }
    return (
      <div className={className} onClick={() => this.changePage(this.props.page + 1)}>Right</div>
    )
  }

  renderPageNumbers() {
    const page_numbers = []
    for (let i=1; i<=this.totalPages(); i++) {
      if (i === this.props.page) {
        page_numbers.push(
          <div className="pagination__button current_page" key={i}>
            <span>{i}</span>
          </div>
        )
      } else {
        page_numbers.push(
          <div className="pagination__button" key={i} onClick={() => this.changePage(i)}>
            <span>{i}</span>
          </div>
        )
      }
    }
    return page_numbers
  }

  render() {
    const { className } = this.props
    return (
      <div className={`${className} pagination`} >
        {this.renderLeft()}
        {this.renderPageNumbers()}
        {this.renderRight()}
      </div>
    )
  }

}
export default decorate(
  styled`
  padding: 10px 10px 0;
  text-align: center;

  .pagination__button {
    border-radius: 5px 5px 0 0;
    font-size: .9rem;
    font-weight: 600;
    display: inline-block;
    opacity: .6;
    padding: 16px;
    cursor: pointer;
  }

  .current_page {
    opacity: 1;
    span {
      border-bottom: 2px solid  ${t(`text`)};
    }
  }
`,
  PaginationControls)
