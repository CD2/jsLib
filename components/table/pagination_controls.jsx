import React from "react"
import PropTypes from "prop-types"
import { withRouter } from "react-router-dom"
import { observer } from "mobx-react"
import { computed } from "mobx"

import { styled } from "lib/utils/theme"
import Button from "lib/components/button"

@styled`

  .pagination__button {
    opacity: .6;
    display: inline-block;
    vertical-align: middle;
  }

  .current_page {
    opacity: 1;
  }
`
@withRouter
@observer
export class PaginationControls extends React.Component {
  static propTypes = {
    className: PropTypes.string,
    history: PropTypes.object,
    location: PropTypes.object,
    onPageChange: PropTypes.func.isRequired,
    page: PropTypes.number.isRequired,
    page_button_limit: PropTypes.number,
    per_page: PropTypes.number.isRequired,
    storePageName: PropTypes.string,
    style: PropTypes.object,
    total_items: PropTypes.number.isRequired,
  }

  static defaultProps = {
    page_button_limit: 6,
    storePageName: null,
  }

  changePage(page_number) {
    if (page_number > 0 && page_number <= this.totalPages()) {
      const { storePageName, location } = this.props
      if (storePageName) {
        this.props.history.replace({
          state: {
            ...location.state,
            [`${storePageName}PageNumber`]: page_number,
          },
        })
      }
      this.props.onPageChange(page_number)
    }
  }

  @computed
  get page() {
    const { storePageName, page } = this.props
    const locationPage =
      this.props.location.state &&
      this.props.location.state[`${storePageName}PageNumber`]

    return locationPage || page
  }

  totalPages() {
    const { total_items, per_page } = this.props
    return Math.ceil(total_items / per_page)
  }

  firstPage() {
    return this.page === 1
  }

  lastPage() {
    return this.page === this.totalPages()
  }

  renderLeft() {
    let className = `pagination__button`
    if (this.firstPage()) className += ` pagination__button--disabled`

    return (
      <Button
        buttonStyle="pagination"
        className={className}
        onClick={() => this.changePage(this.page - 1)}
      >
        Left
      </Button>
    )
  }

  renderRight() {
    let className = `pagination__button`
    if (this.lastPage()) className += ` pagination__button--disabled`
    return (
      <Button
        buttonStyle="pagination"
        className={className}
        onClick={() => this.changePage(this.page + 1)}
      >
        Right
      </Button>
    )
  }

  renderPageNumbers() {
    const page_numbers = []
    const current = this.page
    const visibility = 2
    const count = this.totalPages()

    const elipsis = <div className="pagination__button">...</div>
    const number = n => (
      <Button
        key={n}
        buttonStyle="pagination"
        className={`pagination__button ${current === n ? `current_page` : ``}`}
        onClick={() => this.changePage(n)}
      >
        {n}
      </Button>
    )

    if (count < 5 + visibility * 2) {
      for (let i = 1; i <= count; i++) {
        page_numbers.push(number(i))
      }
      return page_numbers
    }

    page_numbers.push(number(1))
    const min = Math.max(current - visibility, 2)
    if (min > 2) page_numbers.push(elipsis)
    const max = Math.min(
      Math.max(3 + visibility * 2, min + visibility * 2),
      count - 1,
    )
    for (let i = Math.min(min, count - 2 - 2 * visibility); i <= max; i++) {
      page_numbers.push(number(i))
    }
    if (max < count - 1) page_numbers.push(elipsis)
    page_numbers.push(number(count))

    return page_numbers
  }

  render() {
    if (this.props.total_items < this.props.per_page) return null
    return (
      <div className={this.props.className} style={this.props.style}>
        {this.renderLeft()}
        {this.renderPageNumbers()}
        {this.renderRight()}
      </div>
    )
  }
}

// 1  2  3  4  5  6  7  - 70  on < 5
// 1  -  3  4  5  6  7  - 70  on 5
// 1  - 64 65 66 67 68  - 70  on 66
// 1  - 64 65 66 67 68 69 70  on > 66

export default PaginationControls
