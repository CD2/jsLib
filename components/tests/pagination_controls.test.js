import t from 'utils/test_helper'

import { PaginationControls } from '../pagination_controls'

t.describe_component(PaginationControls, { page:1, per_page:2, total_records:10, onPageChange:() => {} }, (c) => {
  describe(`propTypes`, () => {
    c.accepts_prop(`page`)
    c.accepts_prop(`per_page`)
    c.accepts_prop(`total_records`)
    c.accepts_prop(`onPageChange`)
  })

  describe(`functions`, () => {
    c.call_function(`changePage`, 10)
    c.call_function(`totalPages`).expect_return_to_eq(5)
    describe(`first page`, () => {
      c.call_function(`firstPage`).expect_return_to_eq(true)
      c.with_props({ 'page': 2 }, (c) => {
        c.call_function(`firstPage`).expect_return_to_eq(false)
      })
    })
    describe(`lastPage`, () => {
      c.with_props({ 'page': 5 }, (c) => {
        c.call_function(`lastPage`).expect_return_to_eq(true)
      })
      c.with_props({ 'page': 2 }, (c) => {
        c.call_function(`lastPage`).expect_return_to_eq(false)
      })
    })
  })

  describe(`render`, () => {
    describe(`Left`, () => {
      c.with_props({ 'page': 2 }, (c) => {
        c.has_css(`div`).with_text(`‹ Left`)
      })
    })
    describe(`Right`, () => {
      c.has_css(`div`).with_text(`Right ›`)
    })

    describe(`PageNumbers`, () => {
      c.has_css(`div`).with_text(`1`)
      c.has_css(`div`).with_text(`2`)
      c.has_css(`div`).with_text(`3`)
      c.has_css(`div`).with_text(`4`)
      c.has_css(`div`).with_text(`5`)
    })
  })

})
