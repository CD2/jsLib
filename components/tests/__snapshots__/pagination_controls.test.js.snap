// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component <PaginationControls /> functions first page Component <PaginationControls /> renders correctly 1`] = `"<div><div class=\\"pagination__control\\">‹ Left</div><div class=\\"pagination__button\\">1</div><div class=\\"pagination__button current_page\\">2</div><div class=\\"pagination__button\\">3</div><div class=\\"pagination__button\\">4</div><div class=\\"pagination__button\\">5</div><div class=\\"pagination__control\\">Right ›</div></div>"`;

exports[`Component <PaginationControls /> functions lastPage Component <PaginationControls /> renders correctly 1`] = `"<div><div class=\\"pagination__control\\">‹ Left</div><div class=\\"pagination__button\\">1</div><div class=\\"pagination__button\\">2</div><div class=\\"pagination__button\\">3</div><div class=\\"pagination__button\\">4</div><div class=\\"pagination__button current_page\\">5</div><div class=\\"pagination__control disabled\\">Right ›</div></div>"`;

exports[`Component <PaginationControls /> functions lastPage Component <PaginationControls /> renders correctly 2`] = `"<div><div class=\\"pagination__control\\">‹ Left</div><div class=\\"pagination__button\\">1</div><div class=\\"pagination__button current_page\\">2</div><div class=\\"pagination__button\\">3</div><div class=\\"pagination__button\\">4</div><div class=\\"pagination__button\\">5</div><div class=\\"pagination__control\\">Right ›</div></div>"`;

exports[`Component <PaginationControls /> render Left Component <PaginationControls /> renders correctly 1`] = `"<div><div class=\\"pagination__control\\">‹ Left</div><div class=\\"pagination__button\\">1</div><div class=\\"pagination__button current_page\\">2</div><div class=\\"pagination__button\\">3</div><div class=\\"pagination__button\\">4</div><div class=\\"pagination__button\\">5</div><div class=\\"pagination__control\\">Right ›</div></div>"`;

exports[`Component <PaginationControls /> renders correctly 1`] = `"<div><div class=\\"pagination__control disabled\\">‹ Left</div><div class=\\"pagination__button current_page\\">1</div><div class=\\"pagination__button\\">2</div><div class=\\"pagination__button\\">3</div><div class=\\"pagination__button\\">4</div><div class=\\"pagination__button\\">5</div><div class=\\"pagination__control\\">Right ›</div></div>"`;
